system-tests/lib/cre/environment/environment.go

```diff
@@ -85,367 +74,112 @@ type SetupInput struct {
 	JobSpecFactoryFunctions []cre.JobSpecFn
 	// Deprecated: use Capabilities []cre.InstallableCapability instead
 	CapabilitiesContractFactoryFunctions []cre.CapabilityRegistryConfigFn
+
+	StageGen *stagegen.StageGen
 }
 
-func mustGetAddress(dataStore datastore.MutableDataStore, chainSel uint64, contractType string, version string, qualifier string) string {
-	key := datastore.NewAddressRefKey(
-		chainSel,
-		datastore.ContractType(contractType),
-		semver.MustParse(version),
-		qualifier,
-	)
-	addrRef, err := dataStore.Addresses().Get(key)
-	if err != nil {
-		panic(fmt.Sprintf("Failed to get %s %s (qualifier=%s) address for chain %d: %s", contractType, version, qualifier, chainSel, err.Error()))
+func (s *SetupInput) Validate() error {
+	if s == nil {
+		return pkgerrors.New("input is nil")
+	}
+
+	if len(s.CapabilitiesAwareNodeSets) == 0 {
+		return pkgerrors.New("at least one nodeSet is required")
+	}
+
+	if len(s.BlockchainsInput) == 0 {
+		return pkgerrors.New("at least one blockchain is required")
 	}
-	return addrRef.Address
-}
 
-var stageCount = 7
+	if s.JdInput == nil {
+		return pkgerrors.New("jd input is nil")
+	}
+
+	if s.InfraInput == nil {
+		return pkgerrors.New("infra input is nil")
+	}
+
+	return nil
+}
 
 func SetupTestEnvironment(
 	ctx context.Context,
 	testLogger zerolog.Logger,
 	singleFileLogger logger.Logger,
 	input *SetupInput,
 ) (*SetupOutput, error) {
-	topologyErr := libdon.ValidateTopology(input.CapabilitiesAwareNodeSets, *input.InfraInput)
-	if topologyErr != nil {
-		return nil, pkgerrors.Wrap(topologyErr, "failed to validate topology")
+	// Validations and boostraps
```

Comment:
Side note: the best comments are the ones that provide context that isn't easily discernible from the context -- i.e. the explain why we're doing something, not what we're doing